compiler/rustc_codegen_ssa/src/mir/mod.rs

```diff
@@ -142,6 +151,112 @@ impl<'a, 'tcx, V: CodegenObject> LocalRef<'tcx, V> {
     }
 }
 
+struct FindThreadLocal(bool);
+
+impl<'tcx> Visitor<'tcx> for FindThreadLocal {
+    fn visit_rvalue(&mut self, rvalue: &mir::Rvalue<'tcx>, location: mir::Location) {
+        if let mir::Rvalue::ThreadLocalRef(..) = rvalue {
+            self.0 = true;
+        }
+        self.super_rvalue(rvalue, location);
+    }
+}
+
+struct ReplaceThreadLocal<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    local_start: usize,
+    list: Vec<DefId>,
+}
+
+impl<'tcx> MutVisitor<'tcx> for ReplaceThreadLocal<'tcx> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn visit_rvalue(&mut self, rvalue: &mut mir::Rvalue<'tcx>, location: mir::Location) {
+        if let mir::Rvalue::ThreadLocalRef(def_id) = *rvalue {
+            if self.tcx.is_in_upstream_dylib(def_id.krate) {
+                *rvalue = mir::Rvalue::Use(mir::Operand::Copy(mir::Place {
+                    local: mir::Local::new(self.local_start + self.list.len()),
+                    projection: self.tcx.intern_place_elems(&[]),
+                }));
+                self.list.push(def_id);
+            }
+        }
+        self.super_rvalue(rvalue, location);
+    }
+}
+
+// Convert thread local references to thread local function shims if necessary
+fn convert_tls_rvalues<'tcx>(
```

Comment:
Can't this be done on the fly during codegen of `ThreadLocalRef`? That would save an unnecessary clone and arena allocation. It would also prevent memory usage growing without bounds when a function using `ThreadLocalRef` is generic and instantiated a lot of times.

Reply:
By the way cg_clif also needs this change.

Reply:
I was trying to do this on MIR, but MIR can't call MIR shims. I'm not sure if it can be inserted on the fly as instructions need to be inserted, while I think thread locals are values.

Maybe making thread local references instructions in MIR could be a solution? Support for generic statics might also want that for statics in general.

Reply:
When codegening an Rvalue, a Builder is available, so adding a function call in the middle is possible.

Reply:
Is cg_clif in tree now? And could you point me to the Rvalue lowering? :)

Reply:
It has been for a while. The lowering is at https://github.com/rust-lang/rust/blob/73f40197ecabf77ed59028af61739404eb60dd2e/compiler/rustc_codegen_cranelift/src/base.rs#L555-L558 which calls https://github.com/rust-lang/rust/blob/73f40197ecabf77ed59028af61739404eb60dd2e/compiler/rustc_codegen_cranelift/src/constant.rs#L52-L64